### PR TITLE
doc(autotuning): Model auto-tuning framework design doc.

### PR DESCRIPTION
**Pull request by Antlera**
### What changes were proposed in this pull request?

Add hyperparameters auto-tuning design documentation describing how to add hyperparameter tuning modules to dlrover's elastic framework.

### Why are the changes needed?

This document is a framework description of the entire AUTO TUNING.

### Does this PR introduce any user-facing change?

NO.

### How was this patch tested?

NO. This is a doc.

### CODE REVIEW

File: docs/design/autotuing-design.md [hunk 1]
```diff
@@ -0,0 +1,269 @@
+# Design to Auto-Tuning Hyperparameters with DLRover
+
+This design doc describes how DLrover auto-tunes hyperparameters a job.
```
Comment (workingloong) [inline docs/design/autotuing-design.md:3]: "hyperparameters a job." - > "hyperparameters of a job."

File: docs/design/autotuing-design.md [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+up trying things out repeatedly to get the right settings.
+
+## Target
+
+Our primary objective is to design and implement a performance monitoring and
+optimization module for the training process, aimed at enhancing the efficiency
+ and effectiveness of distributed deep learning training. The specific goals
+encompass:
+
+ **Collecting Training Data:** We intend to monitor and gather
```
Comment (workingloong) [inline docs/design/autotuing-design.md:24]: Collecting Training Data ->  Collecting Runtime Metrics.

Training data is usually training datasets.

File: docs/design/autotuing-design.md [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+ **Real-time Analysis of Training Performance Data:** Our
+ aim is to analyze the collected performance data in real-time, allowing us to
+ provide optimization strategies based on the current training status. This
+ includes suggesting new batch sizes and learning rates to maximize GPU
+ resource utilization.
+
+ **Dynamic Batch Size Configuration:** Leveraging the
+ collected and analyzed performance data, we'll devise an algorithm that
+ maximizes GPU resource utilization during task submissions while preventing
+ Out-of-Memory (OOM) errors.
```
Comment (workingloong) [inline docs/design/autotuing-design.md:38]: Here， we can emphasize that our solution does not need to restart the training process.

File: docs/design/autotuing-design.md [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+- **update_batch_size_from_config**
+  - **Parameters**: config_path (String)
+  - **Description**: Reads the batch size information from a configuration file located at `config_path` and updates the current batch size accordingly. This method allows for convenient configuration-based adjustment of the batch size.
+
+```python
+class ElasticDataLoader(DataLoader):
+    def init(self, *args, **kwargs):
+        super(ElasticDataLoader, self).init(*args, **kwargs)
+ self.current_batch_size = self.batch_size
```
Comment (workingloong) [inline docs/design/autotuing-design.md:250]: Indent error.

File: docs/design/autotuing-design.md [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+##### Outputs
+
+- **tuned_hyper_paras**
+  - **Type:** HyperParas
+  - **Description:** Dataclass containing hyper parameters tuned by the hyper-parameter tuner.
+
+
+```python
+def tune_hyper_paras(gpu_stats=[], hyper_para_args=[]):
```
Comment (workingloong) [inline docs/design/autotuing-design.md:186]: tune_hyper_params

File: docs/design/autotuing-design.md [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+    
+    gpu_utilization: float
+    """The GPU utilization as a percentage."""
+```
+
+#### HyperParas
+
+```python
+@dataclass
+class HyperParas:
```
Comment (workingloong) [inline docs/design/autotuing-design.md:111]: We usually use params for short of parameters.

**Review by merlintang — CHANGES_REQUESTED** [review]
for this design, i can not find the difference betwen llm specificed vs other kinds of dl model auto tune? can you list the difference for us to understand why we need to design this component more clearly.

File: docs/design/autotuing-design.md (hunk 1 above)
Comment (Antlera) [reply docs/design/autotuing-design.md]: Fixed by latest commit.

File: docs/design/autotuing-design.md (hunk 2 above)
Comment (Antlera) [reply docs/design/autotuing-design.md]: Tks.I will pay more attention to the expressions. Fixed by latest commit.

File: docs/design/autotuing-design.md (hunk 3 above)
Comment (Antlera) [reply docs/design/autotuing-design.md]: Yes, now we claim our elastic dataloader as real-time configurator in no-stop manner.

File: docs/design/autotuing-design.md (hunk 4 above)
Comment (Antlera) [reply docs/design/autotuing-design.md]: Fixed by latest commit.

File: docs/design/autotuing-design.md (hunk 5 above)
Comment (Antlera) [reply docs/design/autotuing-design.md]: Fixed by latest commit.

File: docs/design/autotuing-design.md (hunk 6 above)
Comment (Antlera) [reply docs/design/autotuing-design.md]: Apologize for my oversight. All of these were fixed by the latest commits.

**Comment by Antlera** [comment]
> for this design, i can not find the difference betwen llm specificed vs other kinds of dl model auto tune? can you list the difference for us to understand why we need to design this component more clearly.

This is a general design for auto-tuning. As for the specified design, model analyzer introduced by #642 will deal with this issue.

The background was updated to illustrate the challenges in hypermeter configuration (both speed-related and model quality related). 
Background:
```
As models grow larger and optimization techniques become more complex, manual
configuration tuning becomes increasingly challenging for users. Larger model
parameters result in extensive activation values and additional GPU memory
cache usage. Even minor batch size adjustments can unexpectedly increase GPU
memory requirements, leading to job crashes due to out-of-memory issues.
Additionally, various methods such as "gradient checkpoints," "ZERO," and
"host-mem offload" have been introduced to mitigate these memory
requirements. However, these optimization techniques introduce a vast array
of configuration options, which can be overwhelming for beginners.
Consequently, configuring training jobs has become highly challenging for
users. This often leads users to resort to repetitive experimentation to find
the optimal settings, resulting in an unsatisfactory training experience due
to extended restart times.
```

**Review by workingloong — APPROVED** [review]

**Review by merlintang — APPROVED** [review]

**Pull request merged**